### DISCUSSION TITLE
Cache not working

### DISCUSSION THREAD

**Discussion by damchey** [spatie/laravel-permission#1780] [Q&A]
Have installed version 3.18 with Laravel 8 and can't seem to get the cache working. Through debugbar I can see that every page makes several queries related to this permissions/roles.

Have used file, redis and database drivers. Haven't changed anything in the config/permission.php file. Using the provided API/methods to create/edit/sync roles & permissions, so not doing it directly through the database.

When I test for cache using `Cache::get('spatie.permission.cache')`, it returns null.

Can confirm that my cache/driver is working by doing `Cache::put('test_name', 'test_value')` I can get the `test_value`.

**Comment by drbyte** [comment]
This package *always* caches the queries for which it is written to do caching (there are some which it doesn't because those wouldn't make sense). There's no turning the cache off or bypassing it ... unless your local environment is configured to use a memory-only cache such as `array`.

**Comment by damchey** [comment]
I believe caching works by default, but I don't see anything in cache. I understand that with the default settings, the below code in `PermisssionRegistrar.php` is caching the permissions with roles in cache with default key `spatie.permission.cache`, but when I check my cache for that key, there is nothing in it. 
```
public function getPermissions(array $params = []): Collection
    {
        if ($this->permissions === null) {
            $this->permissions = $this->cache->remember(self::$cacheKey, self::$cacheExpirationTime, function () {
                return $this->getPermissionClass()
                    ->with('roles')
                    ->get();
            });
        }
```

But if I replace the `self::$cacheKey` with some random text eg: `some.permission`, I can see the permissions in the cache for that key. Have a feeling the cache is being set and then removed in the same page load, doesn't really make sense to save to cache if it can't be retrieved on subsequent page loads.

This is on latest Laravel 8 (8.24). I used the `database` cache drive to test it, also with `redis` I got nothing when looking up cache for `spatie.permission.cache` key.

**Comment by rbarden** [comment]
Using
|software|version|
|-|-|
| laravel-permission | 4.0.1|
| laravel/framework | 8.41 |
| predis/predis | 1.1.7 |

It seems to be solved as the cache is filled.

The only three queries that get run every page load for me (after the cache is warm) are: 
```select column_name as `column_name` from information_schema.columns
where table_schema = 'redacted' and table_name = 'roles'``` and 
```select column_name as `column_name` from information_schema.columns
where table_schema = 'redacted' and table_name = 'permissions'``` which are run when the `Role` and `Permission` models run the `parent::__construct($attributes)` call.

And then: ```select `roles`.*, `model_has_roles`.`model_id` as `pivot_model_id`, `model_has_roles`.`role_id` as `pivot_role_id`, `model_has_roles`.`model_type` as `pivot_model_type` from `roles` inner join `model_has_roles` on `roles`.`id` = `model_has_roles`.`role_id` where `model_has_roles`.`model_id` = 1 and `model_has_roles`.`model_type` = 'App\Models\System\User'``` is run every time to get the user's roles from here: `src/Traits/HasRoles.php:180` 

I think this is all expected behavior.

@damchey have you updated/can you update? If so, can you confirm it's now working for you?

**Comment by damchey** [comment]
Upgraded to **4.0.1** , am still getting 5 queries for every page load.

```
select column_name as `column_name` from information_schema.columns where table_schema = 'vaccine_prod' and table_name = 'permissions'

select column_name as `column_name` from information_schema.columns where table_schema = 'vaccine_prod' and table_name = 'roles'

select `permissions`.*, `model_has_permissions`.`model_id` as `pivot_model_id`, `model_has_permissions`.`permission_id` as `pivot_permission_id`, `model_has_permissions`.`model_type` as `pivot_model_type` from `permissions` inner join `model_has_permissions` on `permissions`.`id` = `model_has_permissions`.`permission_id` where `model_has_permissions`.`model_id` = 1 and `model_has_permissions`.`model_type` = 'App\Models\User'

select `roles`.*, `model_has_roles`.`model_id` as `pivot_model_id`, `model_has_roles`.`role_id` as `pivot_role_id`, `model_has_roles`.`model_type` as `pivot_model_type` from `roles` inner join `model_has_roles` on `roles`.`id` = `model_has_roles`.`role_id` where `model_has_roles`.`model_id` = 1 and `model_has_roles`.`model_type` = 'App\Models\User'

select `permissions`.*, `role_has_permissions`.`role_id` as `pivot_role_id`, `role_has_permissions`.`permission_id` as `pivot_permission_id` from `permissions` inner join `role_has_permissions` on `permissions`.`id` = `role_has_permissions`.`permission_id` where `role_has_permissions`.`role_id` in (1)
```

**Comment by anwarx4u** [comment]
any update?

**Comment by a-ssassi-n** [comment]
Many have raised this issue, still can't find a solution for this!

**Comment by HassanZahirnia** [comment]
Same issues here. In the docs it says there is no need to do any extra steps to enable caching as it is enabled by default. By I don't see any caching in action and I see lots of queries made to DB for the permission/role stuff

**Comment by anwarx4u** [comment]
@freekmurze any comment regarding this issue?

**Comment by arislanhaikal** [comment]
Same issue on my application, using the middleware roles. Too many queries in the DB, so application to be slow. Please fixing this issue!

**Comment by anwarx4u** [comment]
@drbyte any update?###

**Comment by yeaggercom** [comment]
Same issue on my application when listing users. This is a big performance issue.

**Comment by ajuchacko** [comment]
Hi guys,

@damchey, @anwardote, @a-ssassi-n, @xperator, @arislanhaikal, @yeaggercom

cache won't get created - If you create Permissions or Roles also cache won't get created when you assign permissions or role to user

but cache will get created when you actually read

For example:

```
dump(Cache::get('spatie.permission.cache')); // returns null

dump($user->hasPermissionTo('edit-articles')); // creates cache and returns true/false

dd(Cache::get('spatie.permission.cache')); // returns cache
```

the package works exactly as it should be 👍🏻

**Comment by a-ssassi-n** [comment]
> Hi guys,
> 
> @damchey, @anwardote, @a-ssassi-n, @xperator, @arislanhaikal, @yeaggercom
> 
> cache won't get created - If you create Permissions or Roles also cache won't get created when you assign permissions or role to user
> 
> but cache will get created when you actually read
> 
> For example:
> 
> ```
> dump(Cache::get('spatie.permission.cache')); // returns null
> 
> dump($user->hasPermissionTo('edit-articles')); // creates cache and returns true/false
> 
> dd(Cache::get('spatie.permission.cache')); // returns cache
> ```
> 
> the package works exactly as it should be 👍🏻

Have you check the queries with Laravel Debug Bar or any other SQL query viewing tools?

**Comment by ajuchacko** [comment]
Hi @a-ssassi-n,

Yea, I have said that after checking it with ```DB::getQueryLog()``` also this is documented

**Comment by anwarx4u** [comment]
> Hi guys,
> 
> @damchey, @anwardote, @a-ssassi-n, @xperator, @arislanhaikal, @yeaggercom
> 
> cache won't get created - If you create Permissions or Roles also cache won't get created when you assign permissions or role to user
> 
> but cache will get created when you actually read
> 
> For example:
> 
> ```
> dump(Cache::get('spatie.permission.cache')); // returns null
> 
> dump($user->hasPermissionTo('edit-articles')); // creates cache and returns true/false
> 
> dd(Cache::get('spatie.permission.cache')); // returns cache
> ```
> 
> the package works exactly as it should be 👍🏻



![image](https://user-images.githubusercontent.com/21153921/124664735-b0c08100-decd-11eb-83ca-fa1da0595ba4.png)

When I tried to reload any page then always run 5 query. Does it ok? Have there any way to decrease number of queries?

**Comment by HassanZahirnia** [comment]
@ajuchacko91 I don't think we're supposed to directly read from cache. Are we? We're calling functions and methods provided by this library, and I believe it should be implemented inside each one of these functions to try to read from cache whenever possible.

**Comment by ajuchacko** [comment]
@xperator yea correct

When you try to check permission first time on a request package creates the cache and uses it on subsequent calls.
So there will be couple of queries get run but only for the first time

For example:
```
  $user = User::first();

  DB::connection()->enableQueryLog();

  dump($user->hasPermissionTo('edit-articles')); // runs query and creates cache

  dump(DB::getQueryLog()); // will show couple of queries every page load
  
  DB::flushQueryLog(); // flush query log

  dump($user->hasPermissionTo('edit-articles')); // returns true/false according to permission ( gets data from cache)
  
  dump(DB::getQueryLog()); // returns empty array
```

bye

**Comment by GeorgeInteli** [comment]
Yes, cache works. This cache only saves the permissions to roles map (default cache key "spatie.permission.cache")

My question is should we enable per user cache? It could be nice config option to allow caching user permissions.

As for the many queries switching **Spatie\Permission\Models\Role** and **Spatie\Permission\Models\Permission** from **$guarded** to **$fillable** will reduce 2 excess queries on each page load. 

`// protected $guarded = ['id'];`
 `protected $fillable = [
        'name',
        'guard_name'
    ];`

Additionally if there is user cache all queries will be eliminated which may be good use case for some.

**Reply by anwarx4u** [reply]
@GeorgeInteli 
You are right.. this way 2 queries save on each page load. If there does not have any secuirty issue then it should be fixed

**Reply by GeorgeInteli** [reply]
To quote Taylor Otwell he himself suggests fillable over guarded

> As a personal recommendation, I recommend always using $fillable instead of $guarded. In general, $fillable is a safer approach to mass assignment protection because you are forced to list any new columns added to the database to the $fillable array. In contrast, you may easily forget to add a new column to a $guarded array, leaving it open for mass assignment by default.

[Reference](https://blog.laravel.com/security-release-laravel-61835-7240)

**Reply by drbyte** [reply]
Yes, but he's not directly talking about packages there. Whether right or wrong, the choice for using `$guarded` on `id` in this package was specifically to protect the one most sensitive thing, leaving it flexible for you to not have to list all your custom fields if you add any in order for them to be fillable.

You can of course do what you like in your override models.

**Reply by GeorgeInteli** [reply]
Yes, I see your point. I just feel like the use case of someone needing to extend the fields is very low and if that's the case they should extend the base model or replace it (they need custom attributes outside the scope of the package). The defaults that come with the package are sacrificing speed and possibly security which I think are more important.
Either way, this is a package design decision and it is up to you guys to decide.
For now I will just replace the base models on my app level.

P.S. Thank you for the beautiful package btw and have a great weekend everyone :)

**Comment by abanghendri** [comment]
I think I have same question here
![image](https://user-images.githubusercontent.com/36181208/147193393-34a953cb-e82a-492e-a05f-2ff01adf6e4a.png)
this is my debugbar screanshoot, I have 85 permissions item in database, and every page load I got 85 model loaded
it makes my app slower

**Reply by erikn69** [reply]
I have 375 without problems, but 26 querys, why so many querys?
![image](https://user-images.githubusercontent.com/4933954/147248255-d555f3d9-6352-4409-8324-a2db4340ac95.png)

**Comment by alrollo** [comment]
Hi to everybody,

I dont know if my solution is good, but what I'm doing is copy the middleware \Spatie\Permission\Middlewares\PermissionMiddleware::class and change the code to get the user's permissions from cache... then I chankge the link in kernel.php to use my own middleware, in my case I only use permissions from middleware routes...

sorry for my english.
